File: android/app/src/main/java/com/studyapp/domain/model/ProblemSessionRecord.kt
package com.studyapp.domain.model

import kotlinx.serialization.Serializable

@Serializable
enum class ProblemResult {
    CORRECT,
    WRONG,
    REVIEW_CORRECT;

    val title: String
        get() = when (this) {
            CORRECT -> "正解"
            WRONG -> "不正解"
            REVIEW_CORRECT -> "復習正解"
        }
}

@Serializable
data class ProblemSessionRecord(
    val number: Int,
    val result: ProblemResult = ProblemResult.CORRECT,
    val detail: String? = null,
    val subNumber: String? = null
) {
    val normalizedSubNumber: String? get() = subNumber?.trim()?.takeIf { it.isNotEmpty() }
    val stableKey: String get() = normalizedSubNumber?.let { "$number:$it" } ?: number.toString()
    val id: String get() = stableKey
    val displayNumber: String get() = normalizedSubNumber?.let { "${number}問($it)" } ?: "${number}問"

    var isWrong: Boolean
        get() = result == ProblemResult.WRONG
        set(value) {
            // This is a computed property pattern - actual setting done via copy()
        }

    fun withIsWrong(value: Boolean): ProblemSessionRecord =
        copy(result = if (value) ProblemResult.WRONG else ProblemResult.CORRECT)
}

object ProblemSessionReviewResolver {
    fun canonicalInputSession(session: StudySession): StudySession {
        return session.withProblemRecords(canonicalInputRecords(session.problemRecords))
    }

    fun applyingAutomaticReviewCorrect(
        session: StudySession,
        previousResults: MutableMap<String, ProblemResult>
    ): StudySession {
        val resolved = canonicalInputRecords(session.problemRecords).map { record ->
            val inputResult = userInputResult(record.result)
            val resolvedResult = if (
                inputResult == ProblemResult.CORRECT &&
                previousResults[record.stableKey] == ProblemResult.WRONG
            ) {
                ProblemResult.REVIEW_CORRECT
            } else {
                inputResult
            }
            previousResults[record.stableKey] = inputResult
            record.copy(result = resolvedResult)
        }
        return session.withProblemRecords(resolved)
    }

    fun canonicalInputRecords(records: List<ProblemSessionRecord>): List<ProblemSessionRecord> {
        return records
            .map { record ->
                ProblemSessionRecord(
                    number = record.number,
                    result = userInputResult(record.result),
                    detail = record.detail?.trim()?.takeIf { it.isNotEmpty() },
                    subNumber = record.normalizedSubNumber
                )
            }
            .associateBy { it.stableKey }
            .values
            .sortedWith(compareBy<ProblemSessionRecord> { it.number }.thenBy { it.normalizedSubNumber ?: "" })
    }

    private fun userInputResult(result: ProblemResult): ProblemResult {
        return if (result == ProblemResult.WRONG) ProblemResult.WRONG else ProblemResult.CORRECT
    }

    private fun StudySession.withProblemRecords(records: List<ProblemSessionRecord>): StudySession {
        if (records.isEmpty()) {
            return copy(problemRecords = emptyList())
        }
        return copy(
            problemRecords = records,
            problemStart = records.minOf { it.number },
            problemEnd = records.maxOf { it.number },
            wrongProblemCount = records.count { it.result == ProblemResult.WRONG }
        )
    }
}

File: ios/StudyApp/Domain/Models/ProblemSessionRecord.swift
import Foundation

enum StudySessionType: String, Codable, CaseIterable, Hashable {
    case stopwatch = "STOPWATCH"
    case timer = "TIMER"
    case manual = "MANUAL"

    var title: String {
        switch self {
        case .stopwatch: return "ストップウォッチ"
        case .timer: return "タイマー"
        case .manual: return "手動"
        }
    }
}

enum ProblemResult: String, Codable, CaseIterable, Hashable {
    case correct
    case wrong
    case reviewCorrect

    var title: String {
        switch self {
        case .correct: return "正解"
        case .wrong: return "不正解"
        case .reviewCorrect: return "復習正解"
        }
    }
}

struct ProblemSessionRecord: Identifiable, Codable, Hashable {
    var number: Int
    var result: ProblemResult
    var detail: String?
    var subNumber: String?

    var normalizedSubNumber: String? {
        subNumber?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfBlank
    }

    var stableKey: String {
        if let normalizedSubNumber {
            return "\(number):\(normalizedSubNumber)"
        }
        return "\(number)"
    }

    var id: String { stableKey }

    var displayNumber: String {
        if let normalizedSubNumber {
            return "\(number)問(\(normalizedSubNumber))"
        }
        return "\(number)問"
    }

    var isWrong: Bool {
        get { result == .wrong }
        set { result = newValue ? .wrong : .correct }
    }

    init(number: Int, result: ProblemResult, detail: String? = nil, subNumber: String? = nil) {
        self.number = number
        self.result = result
        self.detail = detail
        self.subNumber = subNumber?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfBlank
    }

    init(number: Int, isWrong: Bool, detail: String? = nil, subNumber: String? = nil) {
        self.init(number: number, result: isWrong ? .wrong : .correct, detail: detail, subNumber: subNumber)
    }

    private enum CodingKeys: String, CodingKey {
        case number
        case result
        case isWrong
        case detail
        case subNumber
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        number = try container.decode(Int.self, forKey: .number)
        if let resultValue = try container.decodeIfPresent(String.self, forKey: .result) {
            switch resultValue {
            case ProblemResult.correct.rawValue, "CORRECT":
                result = .correct
            case ProblemResult.wrong.rawValue, "WRONG":
                result = .wrong
            case ProblemResult.reviewCorrect.rawValue, "REVIEW_CORRECT":
                result = .reviewCorrect
            default:
                result = (try container.decodeIfPresent(Bool.self, forKey: .isWrong) ?? false) ? .wrong : .correct
            }
        } else {
            result = (try container.decodeIfPresent(Bool.self, forKey: .isWrong) ?? false) ? .wrong : .correct
        }
        detail = try container.decodeIfPresent(String.self, forKey: .detail)
        subNumber = try container.decodeIfPresent(String.self, forKey: .subNumber)
        subNumber = subNumber?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfBlank
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(number, forKey: .number)
        try container.encode(result, forKey: .result)
        try container.encode(isWrong, forKey: .isWrong)
        try container.encodeIfPresent(detail, forKey: .detail)
        try container.encodeIfPresent(normalizedSubNumber, forKey: .subNumber)
    }
}

enum ProblemSessionReviewResolver {
    static func canonicalInputSession(_ session: StudySession) -> StudySession {
        replacingProblemRecords(in: session, with: canonicalInputRecords(session.problemRecords))
    }

    static func applyingAutomaticReviewCorrect(
        to session: StudySession,
        previousResults: inout [String: ProblemResult]
    ) -> StudySession {
        let resolved = canonicalInputRecords(session.problemRecords).map { record in
            let inputResult = userInputResult(record.result)
            let resolvedResult: ProblemResult = if inputResult == .correct, previousResults[record.stableKey] == .wrong {
                .reviewCorrect
            } else {
                inputResult
            }
            previousResults[record.stableKey] = inputResult
            return ProblemSessionRecord(
                number: record.number,
                result: resolvedResult,
                detail: record.detail,
                subNumber: record.normalizedSubNumber
            )
        }
        return replacingProblemRecords(in: session, with: resolved)
    }

    static func canonicalInputRecords(_ records: [ProblemSessionRecord]) -> [ProblemSessionRecord] {
        var latestByKey = [String: ProblemSessionRecord]()
        for record in records {
            let normalized = ProblemSessionRecord(
                number: record.number,
                result: userInputResult(record.result),
                detail: record.detail?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfBlank,
                subNumber: record.normalizedSubNumber
            )
            latestByKey[normalized.stableKey] = normalized
        }
        return latestByKey.values.sorted { lhs, rhs in
            lhs.number == rhs.number
                ? (lhs.normalizedSubNumber ?? "") < (rhs.normalizedSubNumber ?? "")
                : lhs.number < rhs.number
        }
    }

    private static func userInputResult(_ result: ProblemResult) -> ProblemResult {
        result == .wrong ? .wrong : .correct
    }

    private static func replacingProblemRecords(
        in session: StudySession,
        with records: [ProblemSessionRecord]
    ) -> StudySession {
        guard !records.isEmpty else {
            var updated = session
            updated.problemRecords = []
            return updated
        }
        var updated = session
        updated.problemRecords = records
        updated.problemStart = records.map(\.number).min()
        updated.problemEnd = records.map(\.number).max()
        updated.wrongProblemCount = records.filter(\.isWrong).count
        return updated
    }
}
